g2o/core/base_binary_edge.hpp

```diff
@@ -158,56 +160,119 @@ void BaseBinaryEdge<D, E, VertexXiType, VertexXjType>::linearizeOplus()
   ErrorVector errorBak;
   ErrorVector errorBeforeNumeric = _error;
 
+  // A statically allocated array is far and away the most efficient
+  // way to construct the perturbation vector for the Jacobian. If the
+  // dimension is known at compile time, use directly. If the
+  // dimension is known at run time and is less than 12, use an
+  // allocated array of up to 12. Otherwise, use a fallback of a
+  // dynamically allocated array. The value of 12 is used because
+  // most vertices have a dimension significantly smaller than this.
+  
   if (iNotFixed) {
     internal::QuadraticFormLock lck(*vi);
     //Xi - estimate the jacobian numerically
-    number_t add_vi[VertexXiType::Dimension] = {};
-
-    // add small step along the unit vector in each dimension
-    for (int d = 0; d < VertexXiType::Dimension; ++d) {
-      vi->push();
-      add_vi[d] = delta;
-      vi->oplus(add_vi);
-      computeError();
-      errorBak = _error;
-      vi->pop();
-      vi->push();
-      add_vi[d] = -delta;
-      vi->oplus(add_vi);
-      computeError();
-      errorBak -= _error;
-      vi->pop();
-      add_vi[d] = 0.0;
-
-      _jacobianOplusXi.col(d) = scalar * errorBak;
-    } // end dimension
+
+    const int vi_dim = G2O_VERTEX_I_DIM;
+    
+    if ((VertexXiType::Dimension >= 0) || (vi_dim <= 12)) {
```

Comment:
I guess, we can keep this in here like this for now. And see how to proceed on the constant_vertex branch to have it in a more compact manner.

Reply:
Sounds good.

Reply:
Sounds good.